[BUG] Xcode install fails on Chef 15

## Describe the Bug 

`xcode` resource fails to install Xcode on Chef 15 due to `gmkdir` missing.

## Error Message  

```bash
STDERR: ERROR: Error installing xcode-install: ERROR: Failed to build gem native extension.
```

## Chef Exception  

```bash
Running handlers:
2205
[2019-07-08T12:12:40-07:00] ERROR: Running exception handlers
2206
Running handlers complete
2207
[2019-07-08T12:12:40-07:00] ERROR: Exception handlers complete
2208
Chef Infra Client failed. 224 resources updated in 11 minutes 59 seconds
2209
[2019-07-08T12:12:40-07:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
2210
[2019-07-08T12:12:40-07:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
2211
[2019-07-08T12:12:40-07:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: xcode[install Xcode 9.4.1] (apex_automation::xcode line 6) had an error: Mixlib::ShellOut::ShellCommandFailed: chef_gem[xcode-install] (/tmp/kitchen/cache/cookbooks/macos/resources/xcode.rb line 12) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
2212
---- Begin output of /opt/chef/embedded/bin/gem install xcode-install -q --no-document -v "2.5.0" --source=https://www.rubygems.org --no-document --no-user-install ----
2213
STDOUT: Successfully installed claide-1.0.2
2214
Successfully installed slack-notifier-2.3.2
2215
Successfully installed atomos-0.1.3
2216
Successfully installed CFPropertyList-3.0.0
2217
Successfully installed colored2-3.1.2
2218
Successfully installed nanaimo-0.2.6
2219
Successfully installed xcodeproj-1.10.0
2220
Successfully installed rouge-2.0.7
2221
Successfully installed xcpretty-0.3.0
2222
Successfully installed terminal-notifier-2.0.0
2223
Successfully installed terminal-table-1.8.0
2224
Successfully installed public_suffix-2.0.5
2225
Successfully installed multipart-post-2.0.0
2226
Successfully installed word_wrap-1.0.0
2227
Successfully installed tty-cursor-0.7.0
2228
Successfully installed tty-spinner-0.9.1
2229
Successfully installed babosa-1.0.2
2230
Successfully installed colored-1.2
2231
Successfully installed commander-fastlane-4.4.6
2232
Successfully installed excon-0.64.0
2233
Building native extensions. This could take a while...
2234
STDERR: ERROR: Error installing xcode-install:
2235
ERROR: Failed to build gem native extension.
2236
2237
current directory: /opt/chef/embedded/lib/ruby/gems/2.6.0/gems/unf_ext-0.0.7.6/ext/unf_ext
2238
/opt/chef/embedded/bin/ruby -I /opt/chef/embedded/lib/ruby/site_ruby/2.6.0 -r ./siteconf20190708-27933-15k8xci.rb extconf.rb
2239
checking for -lstdc++... yes
2240
creating Makefile
2241
2242
current directory: /opt/chef/embedded/lib/ruby/gems/2.6.0/gems/unf_ext-0.0.7.6/ext/unf_ext
2243
make "DESTDIR=" clean
2244
2245
current directory: /opt/chef/embedded/lib/ruby/gems/2.6.0/gems/unf_ext-0.0.7.6/ext/unf_ext
2246
make "DESTDIR="
2247
compiling unf.cc
2248
linking shared-object unf_ext.bundle
2249
2250
current directory: /opt/chef/embedded/lib/ruby/gems/2.6.0/gems/unf_ext-0.0.7.6/ext/unf_ext
2251
make "DESTDIR=" install
2252
make: /usr/local/bin/gmkdir: No such file or directory
2253
make: *** [.sitearchdir.time] Error 1
2254
2255
make install failed, exit code 2
```

Related to AB#3477094

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,16 @@
 # Changelog
 
+## [3.1.0] - 2019-08-14
+
+### Added
+
+- Added Chef 15 suites to the matrix. We now support Chef 15.
+
+### Fixed
+
+- Chef 15 introduced an issue with the updated Ruby library where `xcode-install` gem required a gem that needed to be compiled,
+  and failed compilation due to a circular dependency on `gmake` which is only available with Xcode being installed on the system.
+
 ## [3.0.8] - 2019-07-31
 
 ### Fixed
@@ -10,8 +21,8 @@
 Thanks to @jkronborg for these two fixes!
 
 ### Fixed
-- Fixed a guard in `keep_awake` for use on portables. 
-- Fixed incorrect attribute key in the Xcode resource documentation, and added a security suggestion. 
+- Fixed a guard in `keep_awake` for use on portables.
+- Fixed incorrect attribute key in the Xcode resource documentation, and added a security suggestion.
 
 ## [3.0.0] - 2019-02-28
 
@@ -32,7 +43,7 @@ Thanks to @jkronborg for these two fixes!
 - Removed `machine_name` resource along with respective tests and documentation in favor of the `hostname` resource in Chef 14.
 - Removed `xcode` recipe along with respective tests, documentation and node attributes in favor of `command_line_tools` resource which was released in 2.10.0.
 - Removed `disable_software_updates` recipe along with respective tests and documentation in favor of `automatic_software_updates` resource.
-- Removed `default` recipe - it was empty anyway. 
+- Removed `default` recipe - it was empty anyway.
 
 ## [2.10.1] - 2019-01-29
 

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -49,6 +49,8 @@ jobs:
     platforms:
     - high-sierra-chef14
     - mojave-chef14
+    - high-sierra-chef15
+    - mojave-chef15
     suites:
     - default
     - software-updates

--- a/kitchen.yml
+++ b/kitchen.yml
@@ -106,7 +106,6 @@ suites:
   verifier:
     controls:
     - xcode-and-simulators
-    - command-line-tool-sentinel
 
 - name: xcode-from-url
   run_list:
@@ -117,7 +116,6 @@ suites:
   verifier:
     controls:
     - xcode-and-simulators
-    - command-line-tool-sentinel
 
 - name: command-line-tools
   run_list:

--- a/kitchen.yml
+++ b/kitchen.yml
@@ -38,6 +38,27 @@ platforms:
   provisioner:
     product_version: 14
 
+- name: sierra-chef15
+  driver:
+    box: microsoft/macos-sierra
+    box_version: 10.12.6
+  provisioner:
+    product_version: 15
+
+- name: high-sierra-chef15
+  driver:
+    box: microsoft/macos-high-sierra
+    box_version: 10.13.6
+  provisioner:
+    product_version: 15
+
+- name: mojave-chef15
+  driver:
+    box: microsoft/macos-mojave
+    box_version: 10.14.6
+  provisioner:
+    product_version: 15
+
 suites:
 - name: default
   provisioner:

--- a/metadata.rb
+++ b/metadata.rb
@@ -5,7 +5,7 @@ license 'MIT'
 description 'Resources for configuring and provisioning macOS'
 long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
 chef_version '>= 14.0' if respond_to?(:chef_version)
-version '3.0.8'
+version '3.1.0'
 
 source_url 'https://github.com/Microsoft/macos-cookbook'
 issues_url 'https://github.com/Microsoft/macos-cookbook/issues'

--- a/resources/xcode.rb
+++ b/resources/xcode.rb
@@ -7,7 +7,16 @@ property :ios_simulators, Array
 property :download_url, String, default: ''
 
 action :install_gem do
-  command_line_tools 'latest'
+  execute 'install xcode gem' do
+    cwd '/tmp'
+    command <<~BASH
+          curl -sL -O https://github.com/neonichu/ruby-domain_name/releases/download/v0.5.99999999/domain_name-0.5.99999999.gem && \
+          /opt/chef/embedded/bin/gem install --no-document domain_name-0.5.99999999.gem && \
+          /opt/chef/embedded/bin/gem install --no-document --conservative xcode-install && \
+          rm -f domain_name-0.5.99999999.gem
+          BASH
+    not_if { ::File.exist? '/opt/chef/embedded/bin/xcversion' }
+  end
 
   chef_gem 'xcode-install' do
     options('--no-document --no-user-install')

--- a/spec/unit/resources/xcode_spec.rb
+++ b/spec/unit/resources/xcode_spec.rb
@@ -10,23 +10,6 @@ describe 'xcode' do
   before(:each) do
     allow_any_instance_of(MacOS::DeveloperAccount).to receive(:authenticate_with_apple)
       .and_return(true)
-    allow_any_instance_of(MacOS::CommandLineTools).to receive(:macos_version)
-      .and_return('10.13')
-    allow_any_instance_of(MacOS::CommandLineTools).to receive(:softwareupdate_list)
-      .and_return(["Software Update Tool\n",
-                   "\n", "Finding available software\n",
-                   "Software Update found the following new or updated software:\n",
-                   "   * Command Line Tools (macOS El Capitan version 10.11) for Xcode-8.2\n",
-                   "\tCommand Line Tools (macOS El Capitan version 10.11) for Xcode (8.2), 150374K [recommended]\n",
-                   "   * Command Line Tools (macOS High Sierra version 10.13) for Xcode-10.0\n",
-                   "\tCommand Line Tools (macOS High Sierra version 10.13) for Xcode (10.0), 190520K [recommended]\n",
-                   "   * Command Line Tools (macOS Mojave version 10.14) for Xcode-10.0\n",
-                   "\tCommand Line Tools (macOS Mojave version 10.14) for Xcode (10.0), 187321K [recommended]\n",
-                   "   * Command Line Tools (macOS High Sierra version 10.13) for Xcode-9.3\n",
-                   "\tCommand Line Tools (macOS High Sierra version 10.13) for Xcode (9.3), 187312K [recommended]\n",
-                   "   * Command Line Tools (macOS High Sierra version 10.13) for Xcode-9.4\n",
-                   "\tCommand Line Tools (macOS High Sierra version 10.13) for Xcode (9.4), 187380K [recommended]\n"]
-                 )
     allow(MacOS::XCVersion).to receive(:available_versions)
       .and_return(["10\n",
                    "10.1 beta 2\n",

--- a/test/cookbooks/macos_test/recipes/xcode_from_url.rb
+++ b/test/cookbooks/macos_test/recipes/xcode_from_url.rb
@@ -10,4 +10,3 @@ elsif mac_os_x_high_sierra?
     version '10.1'
   end
 end
-

--- a/test/integration/default/controls/command_line_tools_test.rb
+++ b/test/integration/default/controls/command_line_tools_test.rb
@@ -1,0 +1,16 @@
+title 'xcode command line tools'
+
+control 'command-line-tool-sentinel' do
+  title 'Command Line Tools sentinel has been deleted'
+  desc '
+    Verify that the Command Line Tools sentinel has been deleted, and that
+    there are no lingering CLT updates since we should have installed the latest
+  '
+  describe file('/tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress') do
+    it { should_not exist }
+  end
+
+  describe command('/usr/sbin/softwareupdate --list') do
+    its('stdout') { should_not include 'Command Line Tools' }
+  end
+end

--- a/test/integration/default/controls/xcode_test.rb
+++ b/test/integration/default/controls/xcode_test.rb
@@ -10,18 +10,3 @@ control 'xcode-and-simulators' do
     it { should_not be_symlink }
   end
 end
-
-control 'command-line-tool-sentinel' do
-  title 'Command Line Tools sentinel has been deleted'
-  desc '
-    Verify that the Command Line Tools sentinel has been deleted, and that
-    there are no lingering CLT updates since we should have installed the latest
-  '
-  describe file('/tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress') do
-    it { should_not exist }
-  end
-
-  describe command('/usr/sbin/softwareupdate --list') do
-    its('stdout') { should_not include 'Command Line Tools' }
-  end
-end

--- a/test/integration/default/controls/xcode_test.rb
+++ b/test/integration/default/controls/xcode_test.rb
@@ -10,3 +10,18 @@ control 'xcode-and-simulators' do
     it { should_not be_symlink }
   end
 end
+
+control 'command-line-tool-sentinel' do
+  title 'Command Line Tools sentinel has been deleted'
+  desc '
+    Verify that the Command Line Tools sentinel has been deleted, and that
+    there are no lingering CLT updates since we should have installed the latest
+  '
+  describe file('/tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress') do
+    it { should_not exist }
+  end
+
+  describe command('/usr/sbin/softwareupdate --list') do
+    its('stdout') { should_not include 'Command Line Tools' }
+  end
+end

--- a/test/integration/default/inspec.yml
+++ b/test/integration/default/inspec.yml
@@ -5,4 +5,4 @@ copyright: Microsoft Corporation
 copyright_email: chef@microsoft.com
 license: MIT
 summary: Test the macOS cookbook custom resources and recipes
-version: 1.0.1
+version: 1.0.2

